library/waku.nim

```diff
@@ -0,0 +1,399 @@
+
+import
+  std/[sequtils,times,strformat,json,options],
+  strutils,
+  os
+import
+  chronicles,
+  chronos,
+  libp2p/crypto/secp,
+  stew/shims/net,
+  eth/net/nat as todo_delete_this_module
+import
+  ../vendor/nim-libp2p/libp2p/crypto/crypto,
+  ../../waku/common/utils/nat,
+  ../../waku/v2/waku_enr/capabilities,
+  ../../waku/v2/waku_core/message/codec,
+  ../../waku/v2/waku_core/message/message,
+  ../../waku/v2/waku_core/topics/pubsub_topic,
+  ../../waku/v2/node/peer_manager/peer_manager,
+  ../../waku/v2/node/waku_node as waku_node_module,
+  ../../waku/v2/node/builder,
+  ../../waku/v2/node/config,
+  ../../waku/v2/waku_relay/protocol,
+  events/[json_error_event,json_message_event,json_signal_event]
+
+################################################################################
+### Wrapper around the waku node
+################################################################################
+
+################################################################################
+### Exported types
+type
+  ConfigNode* {.exportc.} = object
+    # Struct exported to C
+    host*: cstring
+    port*: uint
+    key*: cstring
+    relay*: bool
+
+### End of exported types
+################################################################################
+
+################################################################################
+### Not-exported components
+
+type
+  EventCallback = proc(signal: cstring) {.cdecl, gcsafe, raises: [Defect].}
+
+var eventCallback:EventCallback = nil
+
+proc relayEventCallback(pubsubTopic: string, data: seq[byte]): Future[void] {.gcsafe, raises: [Defect].} =
+  # Callback that hadles the Waku Relay events. i.e. messages or errors.
+  if eventCallback != nil:
+    let msg = WakuMessage.decode(data)
+    var event: JsonSignal
+    if msg.isOk():
+      event = JsonMessageEvent.new(pubsubTopic, msg.value)
+    else:
+      let errorMsg = string("Error decoding message.") & $msg.error
+      event = JsonErrorEvent.new(errorMsg)
+
+    try:
+      eventCallback(cstring($event))
+    except Exception:
+      error "Exception when calling 'eventCallBack': " &
+            getCurrentExceptionMsg()
+  else:
+    error "eventCallback is nil"
+
+  var retFut = newFuture[void]()
+  retFut.complete()
+  return retFut
+
+proc okResp(message: string): string =
+  $(%* { "result": message })
+
+proc errResp(message: string): string =
+  # Convers an error message into an error-JsonResponse
+  # {
+  #   error: string;
+  # }
+  return $(%* { "error": message })
+
+proc setupNat(natConf, clientId: string, tcpPort, udpPort: Port):
```

Comment:
I must be missing some details here, but why wouldn't [`setupNat`](https://github.com/waku-org/nwaku/blob/b277ce1013563b4e76a3f8694f0eefb1f3528a4d/waku/common/utils/nat.nim#L14) in the `common` utility work?

Reply:
The `setupNat` in the `common` is deprecated with the next comment: `deprecated:
  "Unsafe: this proc quits the app if something is not ok".`
  
 Once this PR is fixed I want to raise a separate PR to fix the `setupNat` proc from `common` and only use it. Also, clean the `eth/net/nat as todo_delete_this_module` :)